[#4] 멀티 모듈 적용

mahmunsen opened the pull request:
### 프로젝트 구조 변경

- 단일 모듈에서 멀티 모듈로 변경하여 프로젝트 구조를 바꿨습니다.

mahmunsen committed:
#1 [chore] 멀티모듈 적용하여 프로젝트 구조 변경

mahmunsen committed:
#0 [docs] .gitignore 수정
- 필요한 설정만 남겼습니다.

mahmunsen committed:
#2 [ci/cd] Jenkinsfile 수정
- indent 간격 조정했습니다.

mahmunsen committed:
#2 [ci/cd] Dockerfile 수정

mahmunsen committed:
#2 [ci/cd] Jenkinsfile 수정
코드 밑부분 수정

mahmunsen committed:
#2 [chore] build.gradle, settings.gradle 수정
코드 밑부분 수정

f-lab-lyan commented on .gitignore:
```diff
@@ -22,3 +16,22 @@
 hs_err_pid*
 replay_pid*
+
+HELP.md
```
이 파일은 민승님 로컬에는 존재하는 것인가요? 아니면 빼주면 좋을 것 같습니다.

mahmunsen replied:
HELP.md, hs_err_pid*, replay_pid* 이 세개 말씀이신가용?

hs_err_pid*
replay_pid*  
위 두 파일은 jvm이 오류가 나서 중단됐을 때 생성되는 로그 파일이라고 알고 있어서 제외하면 좋을 것 같다라고 생각했었는데 빼도 무방한가요..?

f-lab-lyan commented on .gitignore:
```diff
@@ -22,3 +16,22 @@
+.gradle
+build/
+!gradle/wrapper/gradle-wrapper.jar
+!**/src/main/**/build/
+!**/src/test/**/build/
```
이 설정들도 확인 부탁 드릴께요.

mahmunsen replied:
build/
!**/src/main/**/build/
!**/src/test/**/build/

.gitignore에 위와 같이 build 폴더의 내용을 무시하지만 !**/src/main/**/build/, !**/src/test/**/build/ 이 두 개 디렉토리는 무시하지 않는다로 보통 인터넷 찾아보면 기본으로 설정이 되어있는데 저 두 디렉토리도 무시해도 상관없나요..? 뭔가 !로 제외한 이유가 있을 것 같아서 찾아봤는데 잘 안나오는 것 같습니다.

f-lab-lyan commented on .gitignore:
```diff
@@ -22,3 +16,22 @@
+*.ipr
+out/
+!**/src/main/**/out/
+!**/src/test/**/out/
```
IntelliJ 설정들 중에서 지금 현재 민승님께 영향을 미치는 설정만 남기고 필요없는 애들은 지우면 좋을 것 같습니다.

f-lab-lyan commented on Dockerfile:
이 Dockerfile은 이미 빌드가 완료 되었을 때만, 정상적으로 동작할 것 같은데요. 빌드를 해서 jar을 가져올 수 있게 변경해보면 더 좋을 것 같습니다.

mahmunsen replied:
현재 Dockerfile이 빌드된 JAR_FILE을 복사해서 app.jar를 실행하는 것으로 되어있는데 jar를 가져올 수 있게 변경한다는 것은 Dockerfile내에 아래(chatGPT가 보여준 예시코드)와 같이 // Gradle 빌드 수행하는 부분을 넣으라는 말씀이신가요..? 

// 사용할 기본 이미지
FROM openjdk:11

// 소스 코드 복사
COPY . /usr/src/app

// 작업 디렉토리 설정
WORKDIR /usr/src/app

// Gradle 빌드 수행
RUN ./gradlew build

// 빌드된 JAR 파일을 이미지로 복사
ARG JAR_FILE=build/libs/*.jar
COPY ${JAR_FILE} app.jar

// 컨테이너 실행 시 실행될 명령어 설정
ENTRYPOINT ["java", "-jar", "app.jar"]

f-lab-lyan commented on build.gradle:
멀티 프로젝트 구조로 바뀌어도 아래의 build.gradle이 필요한 것일까요? 이렇게 하는 것과 각각의 하위 모듈로 내리는 것의 장단점은 어떤 것이 있나요?

mahmunsen replied:
아래 build.gradle에 넣는 dependency들은 각각 모듈에 통일적으로 적용하는 장점이 있고 각각의 하위 모듈에 있는 build.gradle은 다른 모듈에는 적용하지 않지만 해당 모듈에는 적용해야하는 사항을 넣는 장점이 있는 것 같습니다.

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+    }
+}
+
+plugins {
```
최상위 모듈에 이 플러그인들이 전부 필요한 것인가요? 아니면, 특별히 여기에 넣으면 어떤 이득을 취할 수 있는 것인가요?

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+    version = '0.0.1-SNAPSHOT'
+
+    java {
+        sourceCompatibility = '11'
```
toolchain을 적용하면 몇 가지 이점을 얻을 수 있는데요. 어떤 이점이 있을까요?

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+        // slf4j
+        implementation 'org.slf4j:slf4j-api:1.7.31'
+
+        // Lombok
```
Lombok gradle plugin을 고려해보는건 어떨까요?

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+//    implementation 'com.github.ulisesbocchio:jasypt-spring-boot-starter:3.0.4'
+
+        // slf4j
+        implementation 'org.slf4j:slf4j-api:1.7.31'
```
nit: Check the latest version or check the spring dependency management

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    dependencies {
+//    implementation 'com.github.ulisesbocchio:jasypt-spring-boot-starter:3.0.4'
```
당장 사용하지 않는 dependency는 주석처리가 아니라 지우는 게 코드를 깔끔하게 유지하기 위해서 더 좋을 것 같습니다.

f-lab-lyan commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+        compileOnly 'org.projectlombok:lombok'
+        annotationProcessor 'org.projectlombok:lombok'
+
+        testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
만약 어떤 서브모듈이 spring-boot 모듈이 아닌경우에는 어떻게 해야할까요?

mahmunsen committed:
Merge branch 'f-lab-edu:main' into feature/2_multimodules

f-lab-lyan commented on build.gradle:
```diff
@@ -1,53 +1,66 @@
-}
+    dependencyManagement {
+        imports {
+            mavenBom "org.springframework.cloud:spring-cloud-dependencies:2021.0.5"
```
지금 스프링 클라우드 모듈을 사용하고 있는 곳이 있나요?

f-lab-lyan commented on settings.gradle:
음 ... 몇 가지 우려 사항이 있는데요.
- 각각의 모듈은 어떤 기준으로 나눈 것인지
- 모든 모듈이 지금 꼭 필요한 것인지

일반적으로 common모듈은 꼭 필요한 경우에 그때 만드는 것이 좋을 것 같은데요. 어떻게 생각하세요? 그리고 당장 필요없는 모듈을 미리 만들고 시작하는 것은 over-engineering이 될 가능성이 있어보입니다.

f-lab-lyan commented on god-care-common/build.gradle:
```diff
@@ -0,0 +1,6 @@
+jar { enabled = true }
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-web'
```
common 프로젝트가 starter-web을 디펜던시로 가진단 얘기는 웹서버가 될 수 있다는 얘기처럼 보이는데요. 꼭 가져야하는 이유가 있는 것인가요?

f-lab-lyan commented on god-care-core/build.gradle:
```diff
@@ -0,0 +1,8 @@
+dependencies {
+    implementation project(":god-care-common")
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+//    implementation "org.springframework.boot:spring-boot-starter-data-jpa"
```
nit: Please remove lines commented

f-lab-lyan reviewed (commented):
몇 가지 코멘트를 추가했습니다.

일반적으로 PR은 최소단위로 작성하는 것이 좋은데, 지금 민승님 PR은 규모가 조금 큰 것 같습니다. 꼭 필요한 코드만 남기고 정리해보는 것이 좋을 것 같다는 생각이 듭니다.

혹시 해당 구조가 꼭 필요한 이유가 있다면, 의견 주시면 좋을 것 같습니다.

mahmunsen committed:
#2 [chore] .gitignore 수정, 아래 코드들 삭제
HELP.md
out/
!**/src/main/**/out/
!**/src/test/**/out/

mahmunsen committed:
#2 [chore] @EnableSwagger2 어노테이션 옮김(ApiApplication.java -> SwaggerConfig)

mahmunsen committed:
#1 [chore] god-care-api의 build.gradle 수정
- 스웨거 옯김(god-care-api -> root)
- test 관련 dependencies 옮김(root -> god-care-api)

mahmunsen committed:
#1 [chore] god-care-core, god-care-external 모듈 삭제

mahmunsen committed:
#1 [chore] root 모듈 build.gradle 수정
- 필요하지 않은 dependency 주석 제거
- spring-cloud-dependency 제거
- build 전 설정하는 buildscript 제거

mahmunsen committed:
#3 [feat] 공통 에러 및 응답 코드 작성
- god-care-common에 공통 에러 및 응답 코드 부분에 해당하는 ErrorCode, ApiUtils, Response, CustomException 클래스 등을 작성하였습니다.

mahmunsen committed:
#1 [chore] 각 모듈에서 build.gradle 수정
- root: 호환성 위해서 로그용 slf4j 버전 다운
- god-care-api: external 모듈 제거, 테스트 위한mockito-core dependency 추가
- god-care-core: product 클래스 담을 core 모듈 재생성, god-care-common에 의존성
- settings.gradle: external 모듈 제거

mahmunsen committed:
#3 [feat] ApiUtils 파일 수정
- 안쓰이는 success 메소드 제거
- ResponseEntity 형태로 반환되게끔 수정

mahmunsen committed:
#1 [chore] ApiApplication 클래스 수정
- scanBasePackages에서 external 모듈 디렉토리 제거

mahmunsen committed:
#4 [feat] ProductController에서 상품 등록 api 기능 초안 개발
- ResisterProductRequest(DTO), Product 클래스 필드 2개만 추가하여 간략 개발

mahmunsen committed:
#4 [test] ProductController의 상품 등록 기능에 대한 테스트 코드 작성
- MockMvc 사용, perform으로 http 요청
- repository 사용 대신 HashMap 사용

mahmunsen committed:
Revert "#4 [test] ProductController의 상품 등록 기능에 대한 테스트 코드 작성"

This reverts commit 9a717ace3f2a15bd9aa88978ff64f8ededcba869.

mahmunsen committed:
Revert "#3 [feat] 공통 에러 및 응답 코드 작성"

This reverts commit 3676e2980a756e339fc3851d2df5d1c906d738b3.

mahmunsen committed:
Revert "#3 [feat] ApiUtils 파일 수정"

This reverts commit 4771eba26e8569e8fa2a9947feb975d7d2e93582.

mahmunsen committed:
Revert "#1 [chore] 각 모듈에서 build.gradle 수정"

This reverts commit f62af9c3bc4418b2ea1531a192caab3991425c20.

mahmunsen committed:
Revert "#1 [chore] ApiApplication 클래스 수정"

This reverts commit 1f0f6dcfca0aa2e4b902f8c4d0230a22ca1c7264.

mahmunsen committed:
Revert "#4 [feat] ProductController에서 상품 등록 api 기능 초안 개발"

This reverts commit 4a8fc244d3622e948bab97d4eae9c8e02e9b276c.

mahmunsen committed:
Revert "#2 [chore] @EnableSwagger2 어노테이션 옮김(ApiApplication.java -> SwaggerConfig)"

This reverts commit 2cfc6363bfc1ce4014e4f2555094cec157a01f4a.

mahmunsen committed:
Revert "#1 [chore] root 모듈 build.gradle 수정"

This reverts commit 4fa4322cf697a36067f0fa8755a2843cedab4229.

mahmunsen committed:
Revert "#1 [chore] god-care-api의 build.gradle 수정"

This reverts commit 676de33935d2e3913b4a5ca1a57a709f3260eeb2.

mahmunsen committed:
Revert "#2 [chore] .gitignore 수정, 아래 코드들 삭제"

This reverts commit a429a3337403ef6090b2c5e0d0b9fd3dd090fc02.

mahmunsen committed:
Revert "#1 [chore] god-care-core, god-care-external 모듈 삭제"

This reverts commit 98c82ce1c10f12354fdfd8c4ab37c5e1a90dab63.

mahmunsen committed:
#0 [chore] .gitignore 수정
- 불필요한 것들 제거

mahmunsen committed:
#1 [chore] 불필요한 서브 모듈 제거

mahmunsen committed:
#1 [chore] api모듈 build.gradle 변경
- root모듈에 있던 설정 대부분 api모듈로 이동
- lombok 의존성들 플러그인으로 대체

f-lab-lyan commented on build.gradle:
```diff
@@ -1,53 +1,5 @@
-    id 'java'
-    id 'org.springframework.boot' version '2.7.15'
-    id 'io.spring.dependency-management' version '1.1.4'
+allprojects {
```
이건 사실 subprojects로 하는게 더 좋을 것 같습니다만, 다른 PR에서 작업하셔도 될 것 같습니다.

f-lab-lyan commented on god-care-api/build.gradle:
```diff
@@ -0,0 +1,41 @@
+    }
+}
+
+dependencies {
```
각각의 dependency를 어떤 순서로 배열할지 고민해보시면 더 좋을 것 같은데, 이건 멘토링 시간에 살짝 얘기해보죠.

f-lab-lyan commented on god-care-api/settings.gradle:
이 파일은 루트에 하나만 있어도 되는게 아닐까요?

f-lab-lyan commented on god-care-api/src/main/java/com/godcare/api/ApiApplication.java:
```diff
@@ -0,0 +1,17 @@
+
+@EnableSwagger2
+@SpringBootApplication(scanBasePackages = {"com.godcare.api", "com.godcare.common"})
+@EntityScan("com.godcare.core")
```
이 설정은 필요한 설정인가요?

f-lab-lyan reviewed (approved):
리뷰 단 몇 부분만 빼고는 괜찮은 것 같습니다. 각각 코멘트 확인해보시고, 다음 PR에서 준비중이면, 그대로 진행하시고, 여기에 고쳐야 할 것 같은 부분이 있다면, 고치고 다시 알려주세요. (다음 PR을 위해서 아주 사소한 부분은 넘어가보죠. :-) )

mahmunsen committed:
#1 [chore] api 모듈 ApiApplication 어노테이션 수정
- @EnableSwagger2 -> SwaggerConfig로 옮김
- @EntityScan제거

mahmunsen committed:
#1 [chore] common모듈 build.gradle 수정
- 스프링 부트 어노테이션 쓸 일이 생길 때 추후 추가 예정

mahmunsen committed:
#1 [chore] build.gradle 수정
- allprojects -> subprojects

mahmunsen committed:
#1 [chore] settings.gradle 수정
- api모듈의 settings.gradle 제거
- root모듈로 해당 내용 이동

mahmunsen committed:
#1 [chore] SwaggerConfig 수정
- 어노테이션 추가

f-lab-lyan commented:
고생하셨어요!!! 이제 머지하셔도 될 것 같네요. :-)
